pkg/bpf/collection.go

```diff
@@ -41,6 +48,135 @@ func LoadCollectionSpec(path string) (*ebpf.CollectionSpec, error) {
 	return spec, nil
 }
 
+func RemoveUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	return removeUnreachableTailcalls(spec)
+}
+
+func removeUnreachableTailcalls(spec *ebpf.CollectionSpec) error {
+	type TailCall struct {
+		referenced bool
+		visited    bool
+		spec       *ebpf.ProgramSpec
+	}
+
+	entrypoints := make([]*ebpf.ProgramSpec, 0)
+	tailcalls := make(map[uint32]*TailCall)
+
+	for _, prog := range spec.Programs {
+		var id, slot uint32
+		// Consider any program that doesn't follow the tailcall naming convention
+		// x/y to be an entrypoint.
+		if _, err := fmt.Sscanf(prog.SectionName, "%d/%v", &id, &slot); err != nil {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		const (
+			cilium_calls_map = 2
+		)
+
+		// Any program that does follow the x/y naming convention but not part
+		// of the cilium_calls map is also considered an entrypoint.
+		if id != cilium_calls_map {
+			entrypoints = append(entrypoints, prog)
+			continue
+		}
+
+		tailcalls[slot] = &TailCall{
+			spec: prog,
+		}
+	}
+
+	// Allow test_cilium_calls_[0-9]+, cilium_calls_[0-9]+, cilium_calls_hostns_[0-9]+, and cilium_calls_netdev_[0-9]+
+	// Compile in func and not in global scope to allow for the names to be updated.
+	callsMapRegex := regexp.MustCompile(fmt.Sprintf(
+		"^(test_%s|%s|%s|%s)?[0-9]+$",
+		callsmap.MapName,
+		callsmap.MapName,
+		callsmap.HostMapName,
+		callsmap.NetdevMapName,
+	))
+
+	// Discover all tailcalls that are reachable from the given program.
+	visit := func(prog *ebpf.ProgramSpec, tailcalls map[uint32]*TailCall) {
+		for i := 0; i < len(prog.Instructions); i++ {
+			// The `tail_call_static` C function is always used to call tail calls when
+			// the map index is known at compile time.
+			// Due to inline ASM this generates the following instructions:
+			//   Mov R1, Rx
+			//   Mov R2, <map>
+			//   Mov R3, <index>
+			//   call tail_call
+
+			// Find the tail call instruction.
+			inst := prog.Instructions[i]
+			if !inst.IsBuiltinCall() || inst.Constant != int64(asm.FnTailCall) {
+				continue
+			}
+
+			// We should at least have 3 instructions before the tail call.
+			if i-3 <= 0 {
+				continue
+			}
+
+			// Check that the previous instruction is a mov of the tail call index.
+			movIdx := prog.Instructions[i-1]
+			if movIdx.OpCode.ALUOp() != asm.Mov || movIdx.Dst != asm.R3 {
+				continue
+			}
+
+			// Check that the instruction before that is the load of the tail call map.
+			movR2 := prog.Instructions[i-2]
+			if movR2.OpCode != asm.LoadImmOp(asm.DWord) || movR2.Src != asm.PseudoMapFD {
+				continue
+			}
+
+			ref := movR2.Reference()
```

Comment:
What if: instead of hardcoding the list of allowable maps, add a prefix in the inline asm. Then make sure that the reference here has that prefix, and otherwise `continue`. If the prefix is present we set `referenced = true` and do `ins.SetReference(strings.TrimPrefix("inline_prefix", ...))`.

That way we can avoid knowledge of what the map name should be here, the only thing that matters is that the tail call is coming in via the tail_call_static macro. The downside is that now we need to modify the eBPF to make it load :(

Reply:
As discussed offline, lets keep this as is for now and improve the situation with BTF on calls maps and a better map pinning scheme.